fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java

```diff
@@ -3505,130 +3497,265 @@ public void replayRecoverPartition(RecoverInfo info) {
+        Map<Long, Integer> tasksPerBackend = new HashMap<>();
+        for (CreateReplicaTask task : tasks) {
+            tasksPerBackend.compute(task.getBackendId(), (k, v) -> (v == null) ? 1 : v + 1);
+        }
+        return Collections.max(tasksPerBackend.values());
+    }
+
+    private void buildPartitionsSequentially(long dbId, OlapTable table, List<Partition> partitions, int numReplicas,
+                                             int numBackends) throws DdlException {
+        // Try to bundle at least 200 CreateReplicaTask's in a single AgentBatchTask.
+        // The number 200 is just an experiment value that seems to work without obvious problems, feel free to
+        // change it if you have a better choice.
+        int avgReplicasPerPartition = numReplicas / partitions.size();
+        int partitionGroupSize = Math.max(1, numBackends * 200 / Math.max(1, avgReplicasPerPartition));
+        for (int i = 0; i < partitions.size(); i += partitionGroupSize) {
+            int endIndex = Math.min(partitions.size(), i + partitionGroupSize);
+            List<CreateReplicaTask> tasks = buildCreateReplicaTasks(dbId, table, partitions.subList(i, endIndex));
+            int indexCount = partitions.get(i).getVisibleMaterializedIndicesCount();
+            int timeout = Config.tablet_create_timeout_second * countMaxTasksPerBackend(tasks);
+            int maxTimeout = (endIndex - i) * Config.max_create_table_timeout_second * indexCount;
             try {
-                ok = countDownLatch.await(timeout, TimeUnit.MILLISECONDS);
-            } catch (InterruptedException e) {
-                LOG.warn("InterruptedException: ", e);
-                ok = false;
+                sendCreateReplicaTasksAndWaitForFinished(tasks, Math.min(timeout, maxTimeout));
+                tasks.clear();
+            } finally {
+                for (CreateReplicaTask task : tasks) {
+                    AgentTaskQueue.removeTask(task.getBackendId(), TTaskType.CREATE, task.getSignature());
+                }
             }
+        }
+    }
 
-            if (!ok || !countDownLatch.getStatus().ok()) {
-                errMsg = "Failed to create partition[" + partitionName + "]. Timeout.";
-                // clear tasks
-                AgentTaskQueue.removeBatchTask(batchTask, TTaskType.CREATE);
-
+    private void buildPartitionsConcurrently(long dbId, OlapTable table, List<Partition> partitions, int numReplicas,
+                                             int numBackends) throws DdlException {
+        int timeout = numReplicas / numBackends * Config.tablet_create_timeout_second;
+        int numIndexes = partitions.stream().mapToInt(Partition::getVisibleMaterializedIndicesCount).sum();
+        int maxTimeout = numIndexes * Config.max_create_table_timeout_second;
+        MarkedCountDownLatch<Long, Long> countDownLatch = new MarkedCountDownLatch<>(numReplicas);
+        Thread t = new Thread(() -> {
```

Comment:
give this thread a name for debugging conveniently

Reply:
And better the try catch all code int the thread and print log in catch if any exception throwed

Reply:
Any exception should be thrown to outer side. cause cache should be cleaned when failed.